Test connection for connectors

lovincyrus opened the pull request:
This PR supports the ability to test the connection for connectors. Closes https://linear.app/rilldata/issue/APP-23/add-test-connection-feature-to-all-connectors-ui and https://linear.app/rilldata/issue/APP-275/add-our-standard-loading-animation-icon-into-the-button-for-test-and

Another notable change is that we now test olap connectors using resource-level reconciliation.

Postgres `ImplementsSQLStore` (happy path)

https://github.com/user-attachments/assets/25bb9e61-e21a-490d-ba6f-a9672e936a49


MySQL `ImplementsSQLStore` (unhappy path)

https://github.com/user-attachments/assets/210fab27-7091-4f8e-a70c-924254456e99


Redshift `ImplementsWarehouse` (unhappy path)

https://github.com/user-attachments/assets/30fbad3c-8192-43ca-a7c3-76f001f0b0b6



**Checklist:**
- [ ] Covered by tests
- [x] Ran it and it works as intended
- [x] Reviewed the diff before requesting a review
- [x] Checked for unhandled edge cases
- [x] Linked the issues it closes
- [ ] Checked if the docs need to be updated. If so, create a separate Linear DOCS issue
- [ ] Intend to cherry-pick into the release branch
- [x] I'm proud of this work!

ericokuma commented:
<!-- linear:isThreadRoot -->

**Snowflake feedback:**

[Rill Developer | snowflake_1.yaml - 4 September 2025](https://www.loom.com/share/4e3142775842472c9ed79abeec23de89?sid=e02f832b-7fa5-4b32-86c3-f78cddf4f47b)

lovincyrus commented:
> **Snowflake feedback:**
> 
> [Rill Developer | snowflake_1.yaml - 4 September 2025](https://www.loom.com/share/4e3142775842472c9ed79abeec23de89?sid=e02f832b-7fa5-4b32-86c3-f78cddf4f47b)

Addressed - Marking `database` as required in Snowflake based on [this check](https://github.com/rilldata/rill/blob/main/runtime/drivers/snowflake/snowflake.go#L156-L158)

ericpgreen2 commented on web-common/src/features/sources/modal/submitAddDataForm.ts:
```diff
@@ -193,6 +191,25 @@ export async function submitAddConnectorForm(
+  } else {
+    // Poll for the non-OLAP connectors resource status
+    // If the resource status is not reconciled, rollback the changes
+    const result = await pollConnectorResource(
+      instanceId,
+      connector.name as string,
+    );
```
It looks like this `pollConnectorResource` function includes logic that's redundant to what exists above in this `submitAddConnectorForm` function. 

For example:
- `runtimeServicePutFileAndWaitForReconciliation` waits for reconciliation of the resource, so we shouldn't need the polling logic in `pollConnectorResource`, right?
- `fileArtifacts.checkFileErrors` checks for resource reconciliation errors, so we additionally shouldn't need to check again in `pollConnectorResource`, right?

ericpgreen2 replied:
In general, as much as possible, we should use the same code path for all connectors (both non-OLAP and OLAP).

ericpgreen2 replied:
As discussed in our check-in, for non-OLAP connectors, I don't think it's necessary to do a "test connection" client-side. Per [Naman's PR](https://github.com/rilldata/rill/pull/7426), the runtime's reconciliation process should be testing the connection server-side. You can look at that PR to see which connectors are covered by his work. For example, here's the Athena logic:

<img width="1740" height="479" alt="image" src="https://github.com/user-attachments/assets/f15d04fc-9f2b-4f83-9453-65b0c97e0fcb" />

lovincyrus replied:
Addressing

lovincyrus replied:
Reverted the calling `ListDatabaseSchemas` API changes. I think we should still [use a polling](https://linear.app/rilldata/issue/APP-23/add-test-connection-feature-to-all-connectors-ui#comment-df1c7978) to check for the server's reconcile status. I tried the approach by just relying on `runtimeServicePutFileAndWaitForReconciliation`. It will continue to process the resource in the server side but it won't reflect the reconcile status changes to the UI.

ericpgreen2 replied:
> Non-OLAP Connectors:
> 1. Create connector file with runtimeServicePutConnectorFileAndWaitForReconciliation
> 2. Wait for reconciliation + connection testing during reconciliation
> 3. Create .env file with runtimeServicePutFileAndWaitForReconciliation

As mentioned in my latest review, we'll need to create the `.env` file _before_ testing the connection.

ericpgreen2 replied:
In general, the `submitAddConnectorForm` currently requires a very careful read to understand the difference between OLAP connectors and non-OLAP connectors. To strive toward _one_ codepath, we could request that Platform tests the OLAP connectors server-side during reconciliation – just like they do for the non-OLAP connectors.

ericpgreen2 reviewed (changes requested):
I tried to connect a local Postgres instance, and this error makes it feel like something is wrong:

<img width="899" height="684" alt="image" src="https://github.com/user-attachments/assets/0a5676b1-9d68-4c69-a62a-b1d4ef384fc6" />

I believe it's because the code is currently testing the connection _before_ writing the environment variables to the `.env` file.

An e2e test would be helpful to give us confidence whether or not things are working as expected.

royendo reviewed (approved):
Experience pass: 👍 

Have some follow up, will create sep tickets

lovincyrus commented:
> I tried to connect a local Postgres instance, and this error makes it feel like something is wrong:
> 
> <img alt="image" width="899" height="684" src="https://private-user-images.githubusercontent.com/14206386/489722069-0a5676b1-9d68-4c69-a62a-b1d4ef384fc6.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTgxNDkzNDAsIm5iZiI6MTc1ODE0OTA0MCwicGF0aCI6Ii8xNDIwNjM4Ni80ODk3MjIwNjktMGE1Njc2YjEtOWQ2OC00YzY5LWE2MmEtYjFkNGVmMzg0ZmM2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA5MTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwOTE3VDIyNDQwMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJhZmEyMDM3OTk5NDU4ZDllYWI5ZGU3ODE1MWU5ZjljMzQ3NmI5Yzc5ZWI3MjgwM2NjYmMyM2Q5YWE1NzhkM2ImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.zcesx9i4AM6rrD-FgEHklWCymsxjgmwojkalfj2PIA8">
> I believe it's because the code is currently testing the connection _before_ writing the environment variables to the `.env` file.
> 
> An e2e test would be helpful to give us confidence whether or not things are working as expected.

Fixed and added a couple of e2e.
